lib/windows.dart

```diff
@@ -59,13 +61,20 @@ class _WindowsConditionVariable extends ConditionVariable {
     WakeConditionVariable(_impl);
   }
 
+  static const ERROR_TIMEOUT = 0x5b4;
   @override
-  void wait(covariant _WindowsMutex mutex) {
+  void wait(covariant _WindowsMutex mutex, {Duration? timeout}) {
     const infinite = 0xFFFFFFFF;
     const exclusive = 0;
-    if (SleepConditionVariableSRW(_impl, mutex._impl, infinite, exclusive) ==
-        0) {
-      throw StateError('Failed to wait on a condition variable');
+    final result = SleepConditionVariableSRW(_impl, mutex._impl,
+        timeout == null ? infinite : timeout.inMilliseconds, exclusive);
+
+    if (result != 1) {
+      if (GetLastError() == ERROR_TIMEOUT) {
```

Comment:
Sidenote: unfortunately this is not reliable because `GetLastError` can be clobbered with very low probability when returning from `SleepConditionVariableSRW`. Nothing you can do about it now though.

Reply:
how does it get clobbered? GetLastError is maintained on a per thread basis and there are no other calls after SleepConditionVariableSRW.

Reply:
It can get clobbered because: 

- VM can decide to do a GC when returning from `SleepConditionVariableSRW` at which point it will go to runtime to block the thread (and possibly do GC related actions). 
- `GetLastError` (can be) lazily linked, so when you invoke `GetLastError` for the first time execution would go into runtime to perform resolution and clobber the current value.